File: 6200/6236.py
n,m=map(int,input().split())
D=[int(input()) for i in range(n)]
Min,Max=0,sum(D)
k=float('inf')
while Min<=Max:
    mid=(Min+Max)//2
    ans=0;z=0;s=0
    for i in range(n):
        if ans+D[i]<=mid: ans+=D[i]
        else:s=max(s,ans);ans=D[i];z+=1
    z+=1
    if z<=m: k=min(k,max(s,ans));Max=mid-1
    else: Min=mid+1
print(k)

File: 6200/6236.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
int D[100000]={0,};
long long k = 100000*10001;

int main(){
    cin >> n >> m;
    int Min = 0, Max = 0;
    for(int i = 0; i < n; i++){
        cin >> D[i];
        Max+=D[i];
    }
    while(Min<=Max){
        int mid = (Min+Max)/2;
        long long ans = 0, z = 0, s = 0;
        for(int i = 0; i<n; i++){
            if(ans+D[i]<=mid) ans+=D[i];
            else{
                z=max(z,ans);
                ans=D[i];
                s+=1;
            }
        }
        s+=1;
        if(s<=m){
            k=min(k,max(z,ans));
            Max=mid-1;
        }
        else Min=mid+1;
    }
    cout << k << endl;
}